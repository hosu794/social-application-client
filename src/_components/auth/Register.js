import React, { useState, useEffect } from "react";

import { useSelector, useDispatch } from "react-redux";
import { authActions, alertActions, userActions } from "../../_actions";
import { userService } from "../../_services";
import { createNewUser } from "../../_helpers";
import { rearg } from "lodash";

function Register() {
  const [user, setUser] = useState({
    name: "",
    username: "",
    email: "",
    password: "",
    passwordReapet: "",
  });

  const dispatch = useDispatch();

  const isUsernameAvailable = useSelector(
    (state) => state.user.isUsernameAvailable
  );
  const isEmailAvailable = useSelector((state) => state.user.isEmailAvailable);
  const alert = useSelector((state) => state.alert);
  const registering = useSelector((state) => state.registration.registration);
  const { name, password, passwordReapet, username, email } = user;

  const [error, setError] = useState({
    isTrue: false,
    message: "",
  });

  useEffect(() => {
    dispatch(authActions.logout());
  }, []);

  function clearAlerts() {
    setError((error) => ({
      isTrue: false,
      message: "",
    }));
  }

  function arePasswordSame(password, repeatedPassword) {
    const competePassword = password === repeatedPassword;

    console.log("Compete Passwords");
    if (!competePassword) {
      setError((error) => ({
        isTrue: true,
        message: "Password are not same!",
      }));
      return false;
    } else {
      return true;
    }
  }

  function checkEmailAvailability(email) {
    dispatch(userActions.checkEmailAvailability(email));
    if (!isEmailAvailable) {
      setError((error) => ({
        isTrue: true,
        message: "Email is already taken",
      }));
    }
  }

  function checkUsernameAvailability(username) {
    dispatch(userActions.checkUsernameAvailability(username));
    if (!isUsernameAvailable) {
      setError((error) => ({
        isTrue: true,
        message: "Username is already taken",
      }));
    }
  }

  function handleChange(e) {
    const { name, value } = e.target;
    setUser((user) => ({ ...user, [name]: value }));
  }

  function validateUserFields() {
    if (!user.username) {
      setError((error) => ({
        isTrue: true,
        message: "Username is required!",
      }));
    } else if (!user.name) {
      setError((error) => ({
        isTrue: true,
        message: "User is required!",
      }));
    } else if (!user.email) {
      setError((error) => ({
        isTrue: true,
        message: "Email is required!",
      }));
    } else if (!user.password) {
      setError((error) => ({
        isTrue: true,
        message: "Password is required!",
      }));
    } else if (!user.email) {
      setError((error) => ({
        isTrue: true,
        message: "Email is required!",
      }));
    } else if (!user.passwordReapet) {
      setError((error) => ({
        isTrue: true,
        message: "Reapet Password",
      }));
    }
  }

  function handleAlertButton() {
    clearAlerts();
  }

  function handleSubmit(e) {
    e.preventDefault();

    clearAlerts();

    validateUserFields();

    arePasswordSame(password, passwordReapet);

    checkUsernameAvailability(username);
    checkEmailAvailability(email);

    const validateCrudentialsAndCheckIsError =
      username && name && password && passwordReapet;

    if (arePasswordSame && validateCrudentialsAndCheckIsError) {
      clearAlerts();
      dispatch(authActions.register(createNewUser(user)));
    }
  }

  const showError = error.isTrue;

  return (
    <section className="hero is-success is-fullheight">
      <div className="hero-body">
        <form onSubmit={handleSubmit} className="container has-text-centered">
          {showError && (
            <div className="notification is-danger">
              <button onClick={handleAlertButton} className="delete"></button>
              {error.message}
            </div>
          )}
          <div className="field">
            <p className="control has-icons-left has-icons-right">
              <input
                className="input"
                name="username"
                onChange={handleChange}
                value={username}
                type="text"
                placeholder="Username"
              />

              <span className="icon is-small is-left">
                <i className="fas fa-users"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <p className="control has-icons-left has-icons-right">
              <input
                className="input"
                name="name"
                onChange={handleChange}
                value={name}
                type="text"
                placeholder="Name"
              />

              <span className="icon is-small is-left">
                <i className="fas fa-users"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <p className="control has-icons-left has-icons-right">
              <input
                className="input"
                name="email"
                onChange={handleChange}
                value={email}
                type="email"
                placeholder="Email"
              />

              <span className="icon is-small is-left">
                <i className="fas fa-envelope"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <p className="control has-icons-left">
              <input
                className="input"
                name="password"
                onChange={handleChange}
                value={password}
                type="password"
                placeholder="Password"
              />
              <span className="icon is-small is-left">
                <i className="fas fa-lock"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <p className="control has-icons-left">
              <input
                className="input"
                name="passwordReapet"
                onChange={handleChange}
                value={passwordReapet}
                type="password"
                placeholder="Reapet password"
              />
              <span className="icon is-small is-left">
                <i className="fas fa-lock"></i>
              </span>
            </p>
          </div>
          <div className="field">
            <button
              type="submit"
              className="button is-primary is-fullwidth is-light"
            >
              {registering ? "Registering" : "Register"}
            </button>
          </div>
        </form>
      </div>
    </section>
  );
}

export default Register;
