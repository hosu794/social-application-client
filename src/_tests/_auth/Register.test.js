import React from "react";

import configureStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import Register from "../../_components/auth/Register";
import { mount } from "enzyme";
import { wrap } from "lodash";
const mockStore = configureStore([]);

let store = mockStore({
  authentication: {
    loading: false,
    error: false,
  },
  user: {
    isUsernameAvailable: true,
    isEmailAvailable: true,
  },
  registration: {
    registering: false,
  },
});

describe("Test for the Register Component", () => {
  const TestingComponent = () => (
    <MemoryRouter>
      <Provider store={store}>
        <Register />
      </Provider>
    </MemoryRouter>
  );

  const wrapper = mount(<TestingComponent />);

  test("should render component correct", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should component be defined", () => {
    expect(Register).toBeDefined();
  });

  test("should update username input when it changed", () => {
    wrapper
      .find("input")
      .at(0)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "username",
          value: "exampleusername",
        },
      });

    let element = wrapper.find(Register).find("input").at(0).props().value;

    expect(element).toEqual("exampleusername");
  });

  test("should update name input when it changed", () => {
    wrapper
      .find("input")
      .at(1)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "name",
          value: "Random Name",
        },
      });

    let element = wrapper.find(Register).find("input").at(1).props().value;

    expect(element).toEqual("Random Name");
  });

  test("should update email input when it changed", () => {
    wrapper
      .find("input")
      .at(2)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "email",
          value: "joedoe@example.com",
        },
      });

    let element = wrapper.find(Register).find("input").at(2).props().value;

    expect(element).toEqual("joedoe@example.com");
  });

  test("should update password input when it changed", () => {
    wrapper
      .find("input")
      .at(3)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "password",
          value: "randompassword",
        },
      });

    let element = wrapper.find(Register).find("input").at(3).props().value;

    expect(element).toEqual("randompassword");
  });

  test("should update confirmPassword input when it changed", () => {
    wrapper
      .find("input")
      .at(4)
      .simulate("change", {
        persist: () => {},
        target: {
          name: "confirmPassword",
          value: "randomconfirmpassword",
        },
      });

    let element = wrapper.find(Register).find("input").at(4).props().value;

    expect(element).toEqual("randomconfirmpassword");
  });
});
