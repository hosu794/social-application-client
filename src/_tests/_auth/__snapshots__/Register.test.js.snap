// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the Register Component should render component correct 1`] = `ReactWrapper {}`;
